src/transform/FirstParagraphRelocationTransform.js

```diff
@@ -0,0 +1,92 @@
+import Polyfill from './Polyfill'
+
+/**
+ * Determine if paragraph is the one we are interested in.
+ * @param  {!HTMLParagraphElement}  paragraphElement
+ * @return {boolean}
+ */
+const isParagraphGood = paragraphElement => {
+  // Ignore 'coordinates' which are presently hidden. See enwiki 'Bolton Field' and 'Sharya Forest
+  // Museum Railway'. Not counting coordinates towards the min 'good' textContent length heuristic
+  // has dual effect of p's containing only coordinates being rejected, and p's containing
+  // coordinates but also other elements meeting the min 'good' textContent length being accepted.
+  const coordElement = paragraphElement.querySelector('[id="coordinates"]')
+  const coordTextLength = !coordElement ? 0 : coordElement.textContent.length
+
+  // Ensures the paragraph has at least a little text. Otherwise silly things like a empty P or P
+  // which only contains a BR tag will get pulled up. See enwiki 'Hawaii', 'United States',
+  // 'Academy (educational institution)', 'Lovászpatona'
+  const minLength = 50
+  const lengthExceedsMinLength =
+    paragraphElement.textContent.length - coordTextLength >= minLength
+  return lengthExceedsMinLength
+}
+
+/**
+ * Everything we want to move up (everything between `goodParagraphElement` and the next paragraph).
+ * @param  {!HTMLParagraphElement} goodParagraphElement
+ * @return {!Array.<Element>}
+ */
+const getElementsToMove = goodParagraphElement => {
+  let didHitP = false
+  let didHitNextP = false
+  const shouldElementMoveUp = element => { // eslint-disable-line require-jsdoc
+    if (didHitP && element.tagName === 'P') {
+      didHitNextP = true
+    } else if (element.isEqualNode(goodParagraphElement)) {
+      didHitP = true
+    }
+    return didHitP && !didHitNextP
+  }
+  return Array.from(goodParagraphElement.parentNode.childNodes).filter(shouldElementMoveUp)
+}
+
+/**
+ * Locate first 'good' paragraph. We don't want paragraphs from somewhere in the middle of a table,
+ * so only paragraphs which are direct children of `containerID` element are considered. 
+ * @param  {!Document} document
+ * @param  {!string} containerID ID of the section under examination.
+ * @return {?HTMLParagraphElement}
+ */
+const getFirstGoodParagraph = (document, containerID) =>
+  Polyfill.querySelectorAll(document, `#${containerID} > p`).find(isParagraphGood)
+
+/**
+ * Instead of moving the infobox down beneath the first P tag, move the first 'good' looking P tag
+ * (and related elements) up. This ensures some text will appear above infoboxes, tables, images
+ * etc. This method does not do a 'mainpage' check - do so before calling it.
+ * @param  {!Document} document
+ * @param  {!string} containerID ID of the section under examination.
```

Comment:
`containerID` was used vs passing it an element for efficiency reasons. [This](https://github.com/wikimedia/wikimedia-page-library/pull/122/files#diff-a95381ea972d67bccd5e008dddee34ebR52) query selector seeks `p` elements which are direct children of the content block. This is more efficient performed on the document than doing a query selector (`p`) on the content block element. With the latter you have manually check the parent of each `p` to ensure it's the content block; with the former the query selector enforces this relationship for each `p` thanks to `>`.  The upcoming CSS `Selectors Level 4` includes `Relative Selectors` [syntax](https://drafts.csswg.org/selectors-4/#relative) e.g. `myDiv.querySelectorAll(':scope > p')` to let a query selector _on an element_ only return [direct children](https://stackoverflow.com/a/17206138/135557)